File: Data Structures & Algorithms/invert-a-binary-tree/submission-0.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
/*
    left = this.left
    this.left = this.right
    this.right = left
*/
public class Solution {
    public TreeNode InvertTree(TreeNode root) {
        return DFS(root);
    }
    public TreeNode DFS(TreeNode root) {
        if(root == null){
            return root;
        }
        var left = root.left;
        var right = root.right;
        root.left = right;
        root.right = left;
        DFS(root.left);
        DFS(root.right);
        return root;
        //SWAP at node
        //THEN MOVE LEFT AND RIGHT

    }
}

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    int preindex;
    TreeNode root;
    Map<Integer,Integer> indexes;
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        if(preorder.length == 0) return null;
        indexes = new HashMap<>();
        for(int i=0;i<inorder.length;i++){
            indexes.put(inorder[i],i);
        }
        preindex = 0;
        return dfs(preorder, 0, inorder.length - 1);
    }
    public TreeNode dfs(int[] preorder, int left, int right){
        if(left > right) return null;
        int rootVal = preorder[preindex++];
        TreeNode root = new TreeNode(rootVal);
        int mid = indexes.get(rootVal);
        root.left = dfs(preorder,left,mid-1);
        root.right = dfs(preorder,mid+1,right);
        return root;
    }
}
